core/src/avm1.rs

```diff
@@ -1400,24 +1483,74 @@ impl<'gc> Avm1<'gc> {
         Ok(())
     }
 
-    fn action_new_method(&mut self, _context: &mut UpdateContext) -> Result<(), Error> {
-        let _name = self.pop()?.as_string()?;
-        let _object = self.pop()?.as_object()?;
-        let _num_args = self.pop()?.as_i64()?;
-        self.push(Value::Undefined);
-        // TODO(Herschel)
-        Err("Unimplemented action: NewMethod".into())
+    fn action_new_method(&mut self, context: &mut UpdateContext<'_, 'gc, '_>) -> Result<(), Error> {
+        let method_name = self.pop()?;
+        let object = self.pop()?.as_object()?;
+        let num_args = self.pop()?.as_i64()?;
+        let mut args = Vec::new();
+        for _ in 0..num_args {
+            args.push(self.pop()?);
+        }
+
+        let constructor = object
+            .read()
+            .get(&method_name.as_string()?, self, context, object.to_owned())?
+            .resolve(self, context)?
+            .as_object()?;
+        let prototype = constructor
+            .read()
+            .get("prototype", self, context, constructor)?
+            .resolve(self, context)?
+            .as_object()?;
+
+        let this = prototype.read().new(self, context, prototype, &args)?;
+
+        //TODO: What happens if you `ActionNewMethod` without a method name?
+        constructor
+            .read()
+            .call(self, context, this, &args)?
+            .ignore();
+
+        //TODO: The SWF docs are really cagey about what the hell happens with
+        //user defined constructors... do we need stack continuations?!
+        self.push(Value::Object(this));
+
+        Ok(())
     }
 
-    fn action_new_object(&mut self, _context: &mut UpdateContext) -> Result<(), Error> {
-        let _object = self.pop()?.as_string()?;
+    fn action_new_object(&mut self, context: &mut UpdateContext<'_, 'gc, '_>) -> Result<(), Error> {
+        let fn_name = self.pop()?;
         let num_args = self.pop()?.as_i64()?;
+        let mut args = Vec::new();
         for _ in 0..num_args {
-            let _arg = self.pop()?;
+            args.push(self.pop()?);
         }
+
+        let constructor = self
+            .stack_frames
+            .last()
+            .unwrap()
+            .clone()
+            .read()
+            .resolve(fn_name.as_string()?, self, context)?
+            .resolve(self, context)?
+            .as_object()?;
+        let prototype = constructor
+            .read()
+            .get("prototype", self, context, constructor)?
+            .resolve(self, context)?
+            .as_object()?;
+
+        let this = prototype.read().new(self, context, prototype, &args)?;
+
+        constructor
+            .read()
+            .call(self, context, this, &args)?
+            .ignore();
+
         self.push(Value::Undefined);
```

Comment:
Is this correct?

Reply:
Strictly speaking, the line you highlighted is correct: When you call a constructor, it's return value is ignored and the object that was constructed is returned. However, I just noticed the line right below it is pushing the wrong value. I will change it to `Value::Object(this)` as it should be.